packages/eas-cli/src/credentials/ios/api/GraphqlClient.ts

```diff
@@ -0,0 +1,233 @@
+      projectFullName,
+      { appleAppIdentifierId, iosDistributionType }
+    );
+    let iosAppCredentials: IosAppCredentials;
+    if (maybeIosAppCredentials) {
+      iosAppCredentials = maybeIosAppCredentials;
+    } else {
+      const { account } = appLookupParams;
+      const app = await this.getAppAsync(appLookupParams);
+      await IosAppCredentialsMutation.createIosAppCredentialsAsync(
+        { appleTeamId: appleTeam.id },
+        app.id,
+        account.id
+      );
+      iosAppCredentials = nullthrows(
+        await IosAppCredentialsQuery.withBuildCredentialsByAppIdentifierIdAsync(projectFullName, {
+          appleAppIdentifierId,
+          iosDistributionType,
+        })
+      );
+    }
+
+    const iosAppBuildCredentials = iosAppCredentials.iosAppBuildCredentialsArray?.[0];
+    if (!iosAppBuildCredentials) {
+      return await IosAppBuildCredentialsMutation.createIosAppBuildCredentialsAsync(
+        {
+          iosDistributionType,
+          distributionCertificateId: appleDistributionCertificateId,
+          provisioningProfileId: appleProvisioningProfileId,
+        },
+        iosAppCredentials.id
+      );
+    } else {
+      await IosAppBuildCredentialsMutation.setDistributionCertificateAsync(
+        iosAppBuildCredentials.id,
+        appleDistributionCertificateId
+      );
+      return await IosAppBuildCredentialsMutation.setProvisioningProfileAsync(
+        iosAppBuildCredentials.id,
+        appleProvisioningProfileId
+      );
+    }
+  }
+
+  public async createOrGetExistingAppleTeamAsync(
+    { account }: AppLookupParams,
+    { appleTeamIdentifier, appleTeamName }: { appleTeamIdentifier: string; appleTeamName: string }
+  ): Promise<AppleTeam> {
+    const appleTeam = await AppleTeamQuery.byAppleTeamIdentifierAsync(
+      account.id,
+      appleTeamIdentifier
+    );
+    if (appleTeam) {
+      return appleTeam;
+    } else {
+      return await AppleTeamMutation.createAppleTeamAsync(
+        { appleTeamIdentifier, appleTeamName },
+        account.id
+      );
+    }
+  }
+
+  public async createOrGetExistingAppleAppIdentifierAsync(
+    { account, bundleIdentifier }: AppLookupParams,
+    appleTeam: AppleTeam
+  ): Promise<AppleAppIdentifier> {
+    const appleAppIdentifier = await AppleAppIdentifierQuery.byBundleIdentifierAsync(
+      account.name,
+      bundleIdentifier
+    );
+    if (appleAppIdentifier) {
+      return appleAppIdentifier;
+    } else {
+      return await AppleAppIdentifierMutation.createAppleAppIdentifierAsync(
+        { bundleIdentifier, appleTeamId: appleTeam.id },
+        account.id
+      );
+    }
+  }
+
+  public async getDevicesForAppleTeamAsync(
+    { account }: AppLookupParams,
+    { appleTeamIdentifier }: AppleTeam
+  ): Promise<AppleDevice[]> {
+    return await AppleDeviceQuery.getAllByAppleTeamIdentifierAsync(account.id, appleTeamIdentifier);
+  }
```

Comment:
It seems that inlining the GraphQL queries and mutations in these functions would make them easier to work with. Perhaps we were too quick to move all GraphQL into the `src/graphql/queries` and `src/graphql/mutations` folders. After working with this new code organization, it seems forcing all queries to live in the `src/graphql` folder slows us down and adds an extra layer of indirection.

It's safer to change queries when queries are inline because you know exactly where they're used. (Compare that with a case where two or more unrelated pieces of code share the same query – it's easy to accidentally end up overfetching or underfetching data if you add/remove fields.)

Putting all GraphQL in one folder doesn't create a separation of concerns – it creates a separation by technology used. Your `IosGraphqlClient` is a better abstraction. `IosGraphqlClient` includes all the API operations related to iOS credentials – that's what separation of concerns looks like 👍 

Would you mind moving the GraphQL queries and mutations inline?

Reply:
> It seems that inlining the GraphQL queries and mutations in these functions would make them easier to work with.

It depends on code structure if the functionality you are working with is one or two calls it seems easier to inline and a significant part of the cli will work that way, but almost everything related to credentials is highly interconnected, sometimes we need to pass some object through multiple calls, with inlined calls almost everyone would return slightly different type

> it's easy to accidentally end up overfetching or underfetching data if you add/remove fields

I would argue that we need to overfetch sometimes, (the same case as above credentials management), e.g if we need `appleTeamId` we will fetch `appleTeamName`, because we don't want to create two types, moving that inline won't help with that because the motivation to reuse other parts is still there. We can't really do anything about because types would be unmaintainable very quickly.

> Putting all GraphQL in one folder doesn't create a separation of concerns

Actually, you are mixing terms here, separation of concerns is about creating proper abstraction layers and making sure that they are properly separate from each other so e.g. business logic is not mixed with the data access layer. In fact, your solution (inlining) is breaking the separation of concerns because we mix "concern" of creating a query with a concern of logic that is using data returned by that query. 

In this case, I don't think there is any point in citing design patterns/principles here because:
- I'm not aware of any patter principle that dictates directory structure, that's all just a convention that we need to decide on. We can argue whether directory structure should be dictated by domain or by abstraction layer, but neither of those options is objectively bad. The question is which option is easier and/or safer to use.
- In general patterns like that define how code should be structured when we are talking about doing or computing sth, api is a bit outside of those rules because it's not really part of the system, it's sth external, so a lot of rules can't be applied there

Reply:
Generally, I totally agree with @wkozyra95. The original separation made most sense for me. 
We had API calls (and the implementation detail which was the technology used, in this case: GraphQL) separated from the functions consuming the API.

Inlining queries/mutations in functions calling them causes a few problems:
- long function definitions 
- obfuscates the code with the implementation details
- makes reusing queries/mutations really hard 

However, I don't want to discuss about this anymore. I'll go ahead and move credentials queries/mutations to `credentials/ios/api/graphql` as we discussed on Slack.

In the future, when we have more time for refactoring we should definitely consider better and unified structure for GraphQL code. Even though we agreed on moving these outside the global `graphql` directory, I still don't like it. Inlining queries is probably ok for simple use cases (like updates which you described on Slack) but for such a complex objects like credentials it's a really bad idea.